From c781cfcab5e849d52dba059dca24cc4a01733503 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Mon, 8 Sep 2025 13:56:51 +0800
Subject: [PATCH 123/123] mtk: mt76: mt7996: fix wcid index to follow hw rule

According to HW rule, wcid index shold follow below rule
- Odd tid: secondary wcid
- Even tid: primary wcid
Without this patch, cannot enter WA hardware path and leads
to WA CPU bound.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/main.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index 52fb6332..87d3c76f 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -2685,13 +2685,14 @@ mt7996_net_fill_forward_path(struct ieee80211_hw *hw,
 	struct mtk_wed_device *wed = &dev->mt76.mmio.wed;
 	struct mt7996_sta_link *msta_link;
 	struct mt76_vif_link *mlink;
-	u8 dscp = path->mtk_wdma.tid >> 2;
+	u8 tid = mvif->qos_map[path->mtk_wdma.tid >> 2];
+	u8 link_id = (sta->mlo && tid % 2) ? msta->sec_link : msta->deflink_id;
 
-	mlink = rcu_dereference(mvif->mt76.link[msta->deflink_id]);
+	mlink = rcu_dereference(mvif->mt76.link[link_id]);
 	if (!mlink)
 		return -EIO;
 
-	msta_link = rcu_dereference(msta->link[msta->deflink_id]);
+	msta_link = rcu_dereference(msta->link[link_id]);
 	if (!msta_link)
 		return -EIO;
 
@@ -2720,7 +2721,7 @@ mt7996_net_fill_forward_path(struct ieee80211_hw *hw,
 	path->mtk_wdma.bss = mlink->idx;
 	path->mtk_wdma.queue = 0;
 	path->mtk_wdma.wcid = msta_link->wcid.idx;
-	path->mtk_wdma.tid = mvif->qos_map[dscp];
+	path->mtk_wdma.tid = tid;
 
 	if (ieee80211_hw_check(hw, SUPPORTS_AMSDU_IN_AMPDU) &&
 	    mtk_wed_is_amsdu_supported(wed))
-- 
2.45.2

